Content/Items/Haunted/Weapons.EchochainWhip.cs

```diff
@@ -0,0 +1,730 @@
+			Texture2D bloomTex = ModContent.Request<Texture2D>(AssetDirectory.Keys + "GlowAlpha").Value;
+
+			Asset<Texture2D> texture = ModContent.Request<Texture2D>(Texture);
+			Rectangle whipFrame = texture.Frame(1, 5, 0, 0);
+			int height = whipFrame.Height;
+
+			float fadeOut = 1f;
+			if (Projectile.ai[0] > flyTime * 0.4f && Projectile.ai[0] < flyTime * 0.9f)
+				fadeOut *= 1f - (Projectile.ai[0] - flyTime * 0.4f) / (flyTime * 0.5f);
+			else if (Projectile.ai[0] >= flyTime * 0.9f)
+				fadeOut = 0f;
+
+			for (int i = 15; i > 0; i--)
+			{
+				float fade = 1 - (15f - i) / 15f;
+
+				if (i > 0 && i < tipPositions.Count)
+				{
+					whipFrame.Y = height * 4;
+					Color color = Color.Lerp(new Color(20, 135, 15, 0), new Color(100, 200, 10, 0), fade) * 0.5f;
+
+					Main.EntitySpriteDraw(texture.Value, tipPositions[i] - Main.screenPosition, whipFrame, Color.White * 0.15f * fade * fadeOut, tipRotations[i], whipFrame.Size() * 0.5f, Projectile.scale, SpriteEffects.None, 0);
+
+					Main.spriteBatch.Draw(texBlur, tipPositions[i] - Main.screenPosition, null, Color.White with { A = 0 } * 0.15f * fade * fadeOut, tipRotations[i], texBlur.Size() / 2f, Projectile.scale, 0f, 0f);
+
+					Main.spriteBatch.Draw(bloomTex, tipPositions[i] - Main.screenPosition, null, color * fade * fadeOut, 0f, bloomTex.Size() / 2f, 1f * fade, 0f, 0f);
+
+					Main.spriteBatch.Draw(bloomTex, tipPositions[i] - Main.screenPosition, null, color * fade * fadeOut * 0.25f, 0f, bloomTex.Size() / 2f, 1.5f * fade, 0f, 0f);
+
+					Main.spriteBatch.Draw(bloomTex, tipPositions[i] - Main.screenPosition, null, Color.White with { A = 0 } * 0.15f * fade * fadeOut * 0.4f, 0f, bloomTex.Size() / 2f, 0.8f * fade, 0f, 0f);
+
+					if (i < 15 && i + 1 < tipPositions.Count)
+					{
+						var newPosition = Vector2.Lerp(tipPositions[i], tipPositions[i + 1], 0.5f);
+						float newRotation = MathHelper.Lerp(tipRotations[i], tipRotations[i + 1], 0.5f);
+
+						Main.EntitySpriteDraw(texture.Value, newPosition - Main.screenPosition, whipFrame, Color.White * 0.15f * fade * fadeOut, newRotation, whipFrame.Size() * 0.5f, Projectile.scale, SpriteEffects.None, 0);
+
+						Main.spriteBatch.Draw(texBlur, newPosition - Main.screenPosition, null, Color.White with { A = 0 } * 0.15f * fade * fadeOut, newRotation, texBlur.Size() / 2f, Projectile.scale, 0f, 0f);
+
+						Main.spriteBatch.Draw(bloomTex, newPosition - Main.screenPosition, null, color * fade * fadeOut, 0f, bloomTex.Size() / 2f, 1f * fade, 0f, 0f);
+
+						Main.spriteBatch.Draw(bloomTex, newPosition - Main.screenPosition, null, color * fade * fadeOut * 0.15f, 0f, bloomTex.Size() / 2f, 1.5f * fade, 0f, 0f);
+
+						Main.spriteBatch.Draw(bloomTex, newPosition - Main.screenPosition, null, Color.White with { A = 0 } * 0.15f * fade * fadeOut * 0.4f, 0f, bloomTex.Size() / 2f, 0.8f * fade, 0f, 0f);
+					}
+				}
+			}
+		}
+
+		public override void OnHitNPC(NPC target, NPC.HitInfo hit, int damageDone)
```

Comment:
this projectile onhit looks like it could use a syncing too add 

Owner.TryGetModPlayer(out StarlightPlayer sp);
sp.SetHitPacketStatus(true);

to this as well. looks like there's no client-only logic either?